core/src/avm1/object_reference.rs

```diff
@@ -0,0 +1,216 @@
+    /// If this is `None` then we have previously missed the cache, due to the target object being removed and re-created, causing us to fallback to the slow path resolution
+    pub cached_stage_object: GcCell<'gc, Option<GcWeakCell<'gc, StageObjectData<'gc>>>>,
+}
+
+impl<'gc> MovieClipReference<'gc> {
+    pub fn try_from_stage_object(
+        activation: &mut Activation<'_, 'gc>,
+        stage_object: StageObject<'gc>,
+    ) -> Option<Self> {
+        // We can't use as_display_object + as_movie_clip here as we explicitly don't want to convert `SuperObjects`
+        let display_object = stage_object.as_display_object().unwrap();
+        if let DisplayObject::MovieClip(mc) = display_object {
+            let path = MovieClipPath::new_from_path(activation, mc.path());
+            Some(Self {
+                path,
+                cached_stage_object: GcCell::allocate(
+                    activation.context.gc_context,
+                    Some(stage_object.as_weak()),
+                ),
+            })
+        } else if activation.swf_version() <= 5 {
+            let display_object = Self::process_swf5_references(activation, display_object);
+            let stage_object = display_object
+                .object()
+                .coerce_to_object(activation)
+                .as_stage_object()?;
+            let path = MovieClipPath::new_from_path(activation, display_object.path());
+
+            Some(Self {
+                path,
+                cached_stage_object: GcCell::allocate(
+                    activation.context.gc_context,
+                    Some(stage_object.as_weak()),
+                ),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Handle the logic of swfv5 DisplayObjects
+    fn process_swf5_references(
+        activation: &mut Activation<'_, 'gc>,
+        mut display_object: DisplayObject<'gc>,
+    ) -> DisplayObject<'gc> {
+        // In swfv5 paths resolve to the first MovieClip parent if the target isn't a movieclip
+        if activation.swf_version() <= 5 {
+            while display_object.as_movie_clip().is_none() {
+                if let Some(p) = display_object.avm1_parent() {
+                    display_object = p;
+                }
+            }
+        }
+        display_object
+    }
+
+    /// Resolve this reference to an object
+    /// First tuple param indificates if this path came from the cache or not
+    pub fn resolve_reference(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+    ) -> Option<(bool, Object<'gc>, DisplayObject<'gc>)> {
+        let cache = self.cached_stage_object.read();
+        // Check if we have a cache we can use
+        if let Some(cache) = cache.as_ref() {
+            // Check if we can re-use the cached `DisplayObject`, if we can then take this fast path
+            if let Some(mc) = cache.upgrade(activation.context.gc_context) {
```

Comment:
Is it ever possible for an AVM1 movieclip to become unreachable (and get garbage collected) without first being marked as `removed()`? If this is possible, then the behavior will depend on exactly when our garbage collector runs. If it's not possible, can you add a comment?